### PR TITLE
Dockerize Blueoil

### PR DESCRIPTION
**Pull request by odoku**
## Motivation and Context

In order to improve DeLTA-Lite it was necessary to make the Docker image hostable.

## Description

This PR is based on #105, so please do not merge first.

I made a new Dockerfile for GPU and CPU.
And I wrote a README posted on DockerHub.

Once this PR is merged, set the Automated Build in DockerHub so that the Docker image is automatically hosted.
Then it will be available as follows.

```console
docker run --rm -it blueoil/blueoil:latest init
```

## How has this been tested?

I confirmed that the following procedure passed correctly.

[GPU]
```console
git clone --recursive -b dockerize git@github.com:odoku/blueoil.git

cd blueoil
docker build -t blueoil:0.2.1 -f ./docker/Dockerfile-GPU .
mkdir config dataset saved, result

cd dataset
wget https://s3-ap-northeast-1.amazonaws.com/leapmind-public-storage/datasets/cifar.tgz
tar xvf cifar.tgz

cd ..
docker run --rm \
	-it \
	-v $(pwd)/config:/blueoil/config \
	blueoil:0.2.1 init -o /blueoil/config/cifar.yml

docker run --rm \
	-v $(pwd)/config:/blueoil/config \
	-v $(pwd)/dataset:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	blueoil:0.2.1 train -c /blueoil/config/cifar.yml

docker run --rm \
	-v $(pwd)/dataset:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	blueoil:0.2.1 convert -e train_20190116082924 -p save.ckpt-781

docker run --rm \
	-v $(pwd)/dataset:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	-v $(pwd)/result:/blueoil/result \
	blueoil:0.2.1 predict -i /blueoil/dataset/cifar/test/dog -o /blueoil/result -e train_20190116082924 -p save.ckpt-781
```

[CPU]
```console
git clone --recursive -b dockerize git@github.com:odoku/blueoil.git

cd blueoil
docker build -t blueoil:0.2.1-cpu -f ./docker/Dockerfile-CPU .
mkdir config dataset saved, result

cd dataset
wget https://s3-ap-northeast-1.amazonaws.com/leapmind-public-storage/datasets/cifar.tgz
tar xvf cifar.tgz

cd ..
docker run --rm \
	-it \
	-v $(pwd)/config:/blueoil/config \
	blueoil:0.2.1-cpu init -o /blueoil/config/cifar.yml

docker run --rm \
	-v $(pwd)/config:/blueoil/config \
	-v $(pwd)/dataset:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	blueoil:0.2.1-cpu train -c /blueoil/config/cifar.yml

docker run --rm \
	-v $(pwd)/dataset:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	blueoil:0.2.1-cpu convert -e train_20190116082008 -p save.ckpt-781

docker run --rm \
	-v $(pwd)/dataset:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	-v $(pwd)/result:/blueoil/result \
	blueoil:0.2.1-cpu predict -i /blueoil/dataset/cifar/test/dog -o /blueoil/result -e train_20190116082008 -p save.ckpt-781
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [metadata]
 name = blueoil
-version = 0.1
-url = https://leapmind.io/
+version = 0.2.1
+url = https://blue-oil.org
 author = Blueoil development team
 author_email = test@example.com
```
[the remark this answers is not in the document]
Comment (odoku) [reply setup.cfg]: @djKooks Sorry! Please comment about CLI on #105

File: blueoil/blueoil_init.py [hunk 2]
```diff
@@ -187,15 +190,18 @@ def image_size_filter(raw):
...
-    env = Environment(loader=FileSystemLoader('./blueoil/templates', encoding='utf8'))
+def save_config(blueoil_config, output=None):
+    env = Environment(loader=FileSystemLoader(TEMPLATE_DIR, encoding='utf8'))
     tpl = env.get_template('blueoil-config.tpl.yml')
 
     applied = tpl.render(blueoil_config)
-    config_file = blueoil_config['model_name'] + ".yml"
-    with open(config_file, 'w') as fp:
+
+    if not output:
```
[the remark this answers is not in the document]
Comment (odoku) [reply blueoil/blueoil_init.py]: @djKooks Sorry! Please comment about CLI on #105

**Comment by iizukak** [comment]
@odoku almost looks good. DockerHub hosting is nice.

I find  Dockerfile-GPU is the reduced version of the current Dockerfile.
The reducing line block the hosting?
It is not possible to use current Docker file to host?

**Comment by odoku** [comment]
@iizukak 

Sorry I didn't give a detailed explanation.

## PyENV ceased using it because it is not particularly necessary

https://github.com/blue-oil/blueoil/blob/4fa9fc99fb10a1464ca6d5c941bd18462a48d6b4/docker/Dockerfile#L64-L85
https://github.com/blue-oil/blueoil/blob/0b6f28268c4c6519cfe33597c258bf47c7447500/docker/Dockerfile-GPU#L21-L22

## PyCocoTools was acquired from github

https://github.com/blue-oil/blueoil/blob/4fa9fc99fb10a1464ca6d5c941bd18462a48d6b4/docker/Dockerfile#L103-L105
https://github.com/blue-oil/blueoil/blob/0b6f28268c4c6519cfe33597c258bf47c7447500/docker/Dockerfile-GPU#L55-L56

## I removed the library for development

https://github.com/blue-oil/blueoil/blob/4fa9fc99fb10a1464ca6d5c941bd18462a48d6b4/docker/Dockerfile#L107-L113

## I have set environment variables necessary to run Blueoil CLI

https://github.com/blue-oil/blueoil/blob/0b6f28268c4c6519cfe33597c258bf47c7447500/docker/Dockerfile-GPU#L77-L81

**Comment by iizukak** [comment]
@ruimashita Please review.

**Comment by ruimashita** [comment]
@odoku 
> And I wrote a README posted on DockerHub.
Where is the README file ?

Why do you create new `Dockerfile-GPU`? In order to take care mentenability, it is better to not create to unnecessary new file. I hope you to write the README file under `docker` dir, what and why `Dockerfile`, `Dockerfile-CPU`, `Dockerfile-GPU` is.

**Comment by odoku** [comment]
@ruimashita 

I modified existing ones without creating a new Dockerfile.
And I installed PyCocoTools back from installing from local source.

Here is the [Readme](https://gist.github.com/odoku/dbd1ea12cb380d9797baea9309fad539) for DockerHub.

**Comment by ruimashita** [comment]
@odoku 
Thanks 👍 Great !!

**Comment by ruimashita** [comment]
I will verify the GPU docker work well

**Comment by ruimashita** [comment]
ok to test

**Comment by kchygoe** [comment]
Will this be merged soon...?

**Comment by iizukak** [comment]
@ruimashita 
Can you check this PR?

**Comment by ruimashita** [comment]
@iizukak I didn't verify this PR yet, sorry. Need to solve conflicts to verify, and also maybe there will be some conflict to tensorflow update https://github.com/blue-oil/blueoil/pull/202

File: docker/Dockerfile [hunk 3]
```diff
@@ -14,123 +11,80 @@ deb-src http://ftp.jaist.ac.jp/ubuntu/ xenial-updates main restricted universe m
...
-RUN git clone https://github.com/yyuu/pyenv-virtualenv.git $PYENV_ROOT/plugins/pyenv-virtualenv
+# x-compiler
+RUN ln -s /usr/bin/arm-linux-gnueabihf-g++-5 /usr/bin/arm-linux-gnueabihf-g++
 
-# Pyenv and pyenv virtualenv environment settings
-ENV PATH $PYENV_ROOT/shims:$PYENV_ROOT/bin:$PATH
-RUN eval "$(pyenv init -)" && \
-    eval "$(pyenv virtualenv-init -)"
+# Python
+RUN ln -s /usr/bin/python3.6 /usr/bin/python
```
Comment (tsawada) [inline docker/Dockerfile]:
```suggestion
RUN update-alternatives --install /usr/bin/python python /usr/bin/python3.6 10 
```

**Comment by iizukak** [comment]
run test

**Comment by iizukak** [comment]
@kchygoe @hadusam Can you check ?

**Comment by hadusam** [comment]
I verified operations succeeded correctly by running following commands
```
docker run --rm \
    --runtime=nvidia \
    -e CUDA_VISIBLE_DEVICES \
    -it \
    -v $(pwd)/config:/home/blueoil/config \
    blueoil:0.2.1 blueoil init -o /home/blueoil/config/cifar.yml

docker run --rm \
    --runtime=nvidia \
    -e CUDA_VISIBLE_DEVICES \
    -it \
    -v $(pwd)/config:/home/blueoil/config \
    -v $(pwd)/dataset:/home/blueoil/dataset \
    -v $(pwd)/saved:/home/blueoil/saved \
    blueoil:0.2.1 blueoil train -c /home/blueoil/config/cifar.yml

docker run --rm \
    --runtime=nvidia \
    -e CUDA_VISIBLE_DEVICES \
    -it \
    -v $(pwd)/dataset:/home/blueoil/dataset \
    -v $(pwd)/saved:/home/blueoil/saved \
    blueoil:0.2.1 blueoil convert -e train_20190423161419 -p save.ckpt-3125

docker run --rm \
    --runtime=nvidia \
    -e CUDA_VISIBLE_DEVICES \
    -it \
    -v $(pwd)/dataset:/home/blueoil/dataset \
    -v $(pwd)/saved:/home/blueoil/saved \
    -v $(pwd)/result:/home/blueoil/result \
    blueoil:0.2.1 blueoil predict -i /home/blueoil/dataset/cifar/test/dog -o /home/blueoil/result -e train_20190423161419 -p save.ckpt-3125
```
Some changes from test commands of PR comments
* Commands are invalid, so I change as following, ex: `init` -> `blueoil init`
* Path need to change `/blueoil/` to `/home/blueoil/` because some environment variables are set in dockerfile as using `/home/blueoil` (`DATA_DIR=/home/blueoil/dataset`, `DATA_DIR=/home/blueoil/saved`)
* Add option `runtime=nvidia`
* Add option to pass CUDA_VISIBLE_DEVICES
* Add option `-it` to see realtime output

## Comments
* To use `/blueoil/xxx` instead of `/home/blueoil/xxx`, the default environment variables of `DATA_DIR` and `OUTPUT_DIR` should be changed to `/blueoil/xxx`.
* We should use these commands in `blueoil.sh` to maintain, but I think it may be OK to create another issue/PR after this, because this PR is focused on dockerfile changing.

**Comment by hadusam** [comment]
I try to verify `Dockerfile-CPU`, but got an error on converting.:sob:
I am afraid that there is something wrong in my operation, please check @odoku 
CC @iizukak 
## How to reproduce an error
```
git clone --recursive -b dockerize git@github.com:odoku/blueoil.git
cd blueoil
docker build -t blueoil:0.2.1-cpu -f ./docker/Dockerfile-CPU .
mkdir config dataset saved, result

# Use prepared config
cp tests/config/caltech101_classification.yml config/
export DATA_DIR=/home/blueoil

# Run train
docker run --rm \
    -it \
    -e DATA_DIR \
    -v $(pwd)/config:/home/blueoil/config \
    -v $(pwd)/saved:/home/blueoil/saved \
    blueoil:0.2.1-cpu blueoil train -c /home/blueoil/config/caltech101_classification.yml 

# Run convert
docker run --rm \
    -it \
    -e DATA_DIR \
    -v $(pwd)/saved:/home/blueoil/saved \
    blueoil:0.2.1-cpu blueoil convert -e train_20190424064416 -p save.ckpt-27
```
## Error messages
```
masuda@leapmind-server-24:~/Project/tmp/blueoil$ docker run --rm \
>     -it \
>     -e DATA_DIR \
>     -v $(pwd)/saved:/home/blueoil/saved \
>     blueoil:0.2.1-cpu blueoil convert -e train_20190424064416 -p save.ckpt-27

WARNING: The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
If you depend on functionality not listed there, please file an issue.

WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
----------------- config value --------------------
{'BATCH_SIZE': 1,
 'CLASSES': ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'],
 'DATASET': {'AUGMENTOR': [Blur({'min_value': 0, 'max_value': 1})],
             'BATCH_SIZE': 1,
             'DATA_FORMAT': 'NHWC',
             'PRE_PROCESSOR': [Resize({'size': [128, 128]}), DivideBy255({})]},
 'DATASET_CLASS': <class 'abc.DATASET_CLASS'>,
 'DATA_FORMAT': 'NHWC',
 'IMAGE_SIZE': [128, 128],
 'IS_DEBUG': False,
 'IS_DISTRIBUTION': False,
 'IS_PRETRAIN': False,
 'MAX_EPOCHS': 1,
 'NETWORK': {'ACTIVATION_QUANTIZER': <function linear_mid_tread_half_quantizer at 0x7f6ec0f1cc80>,
             'ACTIVATION_QUANTIZER_KWARGS': {'bit': 2, 'max_value': 2},
             'BATCH_SIZE': 1,
             'DATA_FORMAT': 'NHWC',
             'IMAGE_SIZE': [128, 128],
             'LEARNING_RATE_FUNC': None,
             'LEARNING_RATE_KWARGS': None,
             'OPTIMIZER_CLASS': <class 'tensorflow.python.training.adam.AdamOptimizer'>,
             'OPTIMIZER_KWARGS': {'learning_rate': 0.001},
             'WEIGHT_DECAY_RATE': 0.0005,
             'WEIGHT_QUANTIZER': <function binary_mean_scaling_quantizer at 0x7f6ec0f1cd90>,
             'WEIGHT_QUANTIZER_KWARGS': {}},
 'NETWORK_CLASS': <class 'lmnet.networks.classification.lmnet_v1.LmnetV1Quantize'>,
 'POST_PROCESSOR': None,
 'PRETRAIN_DIR': '',
 'PRETRAIN_FILE': '',
 'PRETRAIN_VARS': [],
 'PRE_PROCESSOR': [Resize({'size': [128, 128]}), DivideBy255({})],
 'SAVE_STEPS': 1000,
 'SUMMARISE_STEPS': 100,
 'TASK': <Tasks.CLASSIFICATION: 'IMAGE.CLASSIFICATION'>,
 'TEST_STEPS': 1000}
----------------- config value --------------------
Restore from /home/blueoil/saved/train_20190424064416/checkpoints/save.ckpt-27
WARNING:tensorflow:From /home/blueoil/lmnet/lmnet/blocks.py:92: conv2d (from tensorflow.python.layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.conv2d instead.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/function.py:1007: calling Graph.create_op (from tensorflow.python.framework.ops) with compute_shapes is deprecated and will be removed in a future version.
Instructions for updating:
Shapes are always computed; don't use the compute_shapes as it has no effect.
WARNING:tensorflow:From /home/blueoil/lmnet/lmnet/networks/classification/lmnet_v1.py:86: dropout (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dropout instead.
WARNING:tensorflow:From /home/blueoil/lmnet/lmnet/networks/classification/lmnet_v1.py:107: average_pooling2d (from tensorflow.python.layers.pooling) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.average_pooling2d instead.
WARNING:tensorflow:From /home/blueoil/lmnet/lmnet/networks/classification/base.py:140: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/saver.py:1266: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to check for files with this prefix.
WARNING:tensorflow:From /home/blueoil/lmnet/lmnet/utils/executor.py:94: convert_variables_to_constants (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.convert_variables_to_constants
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/graph_util_impl.py:245: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.extract_sub_graph
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.OutOfRangeError: Node 'conv1/BatchNorm/FusedBatchNorm' (type: '_FusedConv2D', num of outputs: 1) does not have output 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/blueoil", line 11, in <module>
    load_entry_point('blueoil==0.5.0', 'console_scripts', 'blueoil')()
  File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/blueoil/blueoil/cli.py", line 120, in convert
    export_output_root_dir = run_convert(experiment_id, restore_path, template)
  File "/home/blueoil/blueoil/blueoil_convert.py", line 118, in run
    export_dir = run_export(experiment_id, restore_path=restore_path)
  File "/home/blueoil/lmnet/executor/export.py", line 207, in run
    return _export(config, restore_path, image)
  File "/home/blueoil/lmnet/executor/export.py", line 143, in _export
    val = sess.run(op_output.name, feed_dict=feed_dict)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.OutOfRangeError: Node 'conv1/BatchNorm/FusedBatchNorm' (type: '_FusedConv2D', num of outputs: 1) does not have output 1
```

**Comment by hadusam** [comment]
## Additional info of that error
I tried to run convert with GPU docker image to the same checkpoint, it succeeded.
So trained model created by CPU docker image may be correct.
```
docker run --rm \
    -it \
    -e DATA_DIR \
    -v $(pwd)/saved:/home/blueoil/saved \
    blueoil:0.2.1 blueoil convert -e train_20190424064416 -p save.ckpt-27
```

File: docker/Dockerfile [hunk 4]
```diff
@@ -14,122 +11,80 @@ deb-src http://ftp.jaist.ac.jp/ubuntu/ xenial-updates main restricted universe m
...
-ENV PYTHONPATH $PYTHONPATH:/home/blueoil:/home/blueoil/lmnet:/home/blueoil/dlk/python/dlk
+# Locale setting
+RUN locale-gen en_US.UTF-8
+ENV LANG=en_US.UTF-8 \
+    LANGUAGE=en_US:en \
+    LC_ALL=en_US.UTF-8
+
+ENV PYTHONPATH=/home/blueoil:/home/blueoil/lmnet:/home/blueoil/dlk/python/dlk \
+    MPLBACKEND=Agg \
+    DATA_DIR=/home/blueoil/dataset \
```
[the remark this answers is not in the document]
Comment (ruimashita) [reply docker/Dockerfile]: Either is fine 😄

**Comment by kchygoe** [comment]
How does this go on?

**Comment by hadusam** [comment]
https://github.com/tensorflow/tensorflow/issues/21328

**Comment by kchygoe** [comment]
This PR got invalid. Please Create new one.

**Pull request closed without merging**